### PR TITLE
Updated registry integration to utilize v2 functionality.

### PR DESCRIPTION
**Pull request by Randy808**
The `GET /assets/registry` response still includes all the fields before but this PR adds a `registry` property to it. The `registry.rs` file was also changed to pull information from the service instead of from the asset cache made from the git registry.

The added registry property takes the following shape:
```
{
    "asset_id": "<asset-id>",
    "contract": {
      "entity": {
        "domain": "example.com"
      },
      "name": "Registry Asset",
      "precision": 8,
      "ticker": "REG",
      "version": 1,
      "custom_contract_field": "preserved"
    },
    "initial_issuer_pubkey": "<pubkey>",
    "initial_issuer_pubkey_source": "contract",
    "current_issuer_pubkey": "<pubkey>",
    "issuer_pubkey_history": [],
    "mutable": {
      "category_tags": [
        "stablecoin"
      ]
    },
    "admin": {
      "featured": true
    },
    "icon": {
      "href": "https://registry.example/v2/assets/<asset-id>/icon/<hash>.png"
    },
    "status": "active",
    "created_at": "2026-01-01T00:00:00Z",
    "updated_at": "2026-01-02T00:00:00Z"
  }
```

### CODE REVIEW

File: src/elements/registry.rs [hunk 1]
```diff
@@ -1,123 +1,504 @@
...
+use serde::de::DeserializeOwned;
+use serde_json::{Map as JsonMap, Value as JsonValue};
+use tokio::sync::{oneshot, Mutex, OwnedSemaphorePermit, Semaphore};
+use url::Url;
 
 use crate::errors::*;
 
-// length of asset id prefix to use for sub-directory partitioning
-// (in number of hex characters, not bytes)
+const REGISTRY_CONNECT_TIMEOUT: Duration = Duration::from_secs(2);
```
Comment (Randy808) [inline src/elements/registry.rs]: Can we make `REGISTRY_CONNECT_TIMEOUT` and `REGISTRY_REQUEST_TIMEOUT` into configurable env vars (with defaults)?

File: src/elements/registry.rs [hunk 2]
```diff
@@ -1,123 +1,504 @@
...
+        }
+
+        if !base_url.path().ends_with('/') {
+            let path = format!("{}/", base_url.path());
+            base_url.set_path(&path);
+        }
+        base_url.set_query(None);
+        base_url.set_fragment(None);
+
+        let http = Client::builder()
```
Comment (Randy808) [inline src/elements/registry.rs:132]: Can we change our http client to reject redirects?

File: src/elements/registry.rs [hunk 3]
```diff
@@ -1,123 +1,504 @@
...
+
+    pub async fn get_asset(
+        &self,
+        asset_id: &AssetId,
+    ) -> std::result::Result<Option<RegistryAsset>, RegistryError> {
+        let receiver = {
+            let mut cache = self.asset_cache.lock().await;
+            let now = Instant::now();
+
+            if let Some(AssetCacheEntry::Ready { fetched_at, asset }) = cache.get(asset_id) {
```
Comment (Randy808) [inline src/elements/registry.rs]: Can this be moved out of the `receiver` block?

File: src/elements/registry.rs [hunk 4]
```diff
@@ -1,123 +1,504 @@
...
+
+                // Same-key callers join the in-flight request above without consuming a permit.
+                // New requests are rejected instead of accumulating behind the registry.
+                let permit = self.try_acquire_permit()?;
+                let (sender, receiver) = oneshot::channel();
+                cache.insert(*asset_id, AssetCacheEntry::Fetching(vec![sender]));
+
+                let client = self.clone();
+                let asset_id = *asset_id;
+                tokio::spawn(async move {
```
Comment (Randy808) [inline src/elements/registry.rs]: What's the value of `tokio::spawn` here when we await its result with the receiver right after?

Comment (Randy808) [reply src/elements/registry.rs]: `tokio::spawn`ing into another task allows for cache cleanup if the electrs client were to disconnect

File: src/rest.rs [hunk 5]
```diff
@@ -566,6 +570,13 @@ fn spawn_conn(
             if let Some(ref origins) = config.cors {
                 resp.headers_mut()
                     .insert("Access-Control-Allow-Origin", origins.parse().unwrap());
+                #[cfg(feature = "liquid")]
+                resp.headers_mut().insert(
+                    "Access-Control-Expose-Headers",
+                    "X-Asset-Registry-Status, X-Total-Results"
```
Comment (philippem) [inline src/rest.rs]: what needs these custom headers?

**Comment by philippem** [comment]
Some review analysis comments:


  1. contract is no longer a faithful passthrough — this breaks contract-hash verification.

  registry.rs:502 types the contract and from_registry_asset re-serializes it via serde_json::to_value. But the optional fields have #[serde(default)] without skip_serializing_if:

  #[serde(default)] pub ticker: Option<String>,
  #[serde(default)] pub initial_issuer_pubkey: Option<String>,
  #[serde(default)] pub issuer_pubkey: Option<String>,

  An absent ticker round-trips to an explicit "ticker": null. Previously AssetMeta.contract was a raw JsonValue passed through untouched (6a734f0^:src/elements/registry.rs:124).

  A Liquid asset ID commits to the contract bytes. Any client verifying that the asset ID matches the returned contract now computes a different hash for every asset whose contract omits an   
  optional field. AssetMeta's own fields use skip_serializing_if — the pattern just wasn't applied one struct down.

  2. One missing local asset 503s the entire /assets/registry page.

  .ok_or(RegistryError::MissingLocalAsset(asset_id))?

  collected into Result<Vec<_>, _>, so a single registry entry absent from the local index fails the whole listing. The old code had the same shape, but the old asset DB was local and synced  
  alongside — registry and index moved together. Now they're independently versioned services, so this fires whenever the registry is ahead: during initial sync, after a reindex, or on an     
  asset issued minutes ago.

  MissingLocalAsset also maps to 503, which advertises "retry later" for a condition that may never resolve. Skip the entry (or return it registry-only) rather than failing the page.

**Comment by Randy808** [comment]
Updated branch to no longer serialize missing contract fields with null (if not explicitly present and set to null in the registry), and to skip unavailable registry entries instead of returning a 503

File: src/new_index/query.rs [hunk 6]
```diff
@@ -299,31 +317,93 @@ impl Query {
...
+            let metadata = AssetMeta::from_registry_asset(registry_asset)?;
+            match lookup_asset(self, &asset_id, Some(metadata))
+                .map_err(|error| RegistryError::LocalLookup(error.to_string()))?
+            {
+                Some(asset) => results.push(asset),
+                None => warn!(
+                    "registered asset {} is not yet available in the local index",
+                    asset_id
+                ),
+            }
```
Comment (FedOken) [inline src/new_index/query.rs:405]: `lookup_asset` here does synchronous RocksDB work. This runs inline on the tokio worker thread, `list_registry_assets` is `async` and reached from `handle_asset_registry_request` without `spawn_blocking`, and this loop repeats it up to `ASSETS_MAX_PER_PAGE` times per request.

Worth wrapping the local-lookup portion in `spawn_blocking`.

File: src/elements/registry.rs [hunk 7]
```diff
@@ -128,185 +648,847 @@ pub struct AssetMeta {
...
+        return false;
+    }
+    let first = label.bytes().next().unwrap();
+    let last = label.bytes().last().unwrap();
+    first.is_ascii_alphanumeric()
+        && last.is_ascii_alphanumeric()
+        && label
+            .bytes()
+            .all(|byte| byte.is_ascii_alphanumeric() || byte == b'-')
+}
```
Comment (FedOken) [inline src/elements/registry.rs:857]: Can we use `url::Host::parse`?

File: src/new_index/query.rs (hunk 6 above)
Comment (Randy808) [reply src/new_index/query.rs]: Done

File: src/elements/registry.rs (hunk 7 above)
Comment (Randy808) [reply src/elements/registry.rs]: Done

**Review by agoodminute — DISMISSED** [review]

File: src/elements/registry.rs [hunk 8]
```diff
@@ -1,123 +1,643 @@
...
+        let mut cache = self.asset_cache.lock().await;
+        let now = Instant::now();
+
+        if let Some(AssetCacheEntry::Ready { fetched_at, asset }) = cache.get(asset_id) {
+            if now.duration_since(*fetched_at) < self.asset_cache_ttl {
+                return Ok(asset.as_ref().map(|asset| asset.as_ref().clone()));
+            }
+        }
+
+        let receiver = if let Some(AssetCacheEntry::Fetching(waiters)) = cache.get_mut(asset_id) {
```
Comment (EddieHouston) [inline src/elements/registry.rs]: **A `Fetching` entry whose fetch task dies poisons the asset id forever**

The `Fetching(waiters)` entry is only ever replaced by `complete_asset_fetch`. If the task spawned below panics, or is aborted/dropped at runtime shutdown, `complete_asset_fetch` never runs and the entry stays in the map — `prune_asset_cache` retains `Fetching` entries unconditionally, so it is never evicted either.

Every subsequent `get_asset` for that id then pushes a new oneshot sender into the vec (which lives inside the map, so it is never dropped) and awaits a receiver that can never resolve. `receiver.await` has no timeout, and `spawn_conn` only wraps body collection in a timeout — `handle_request` itself is awaited untimed — so each such request hangs its connection indefinitely.

Suggested: store a deadline on `Fetching` and let `prune_asset_cache` evict stale ones, and/or wrap the `receiver.await` in a timeout so a lost notification degrades to one failed request rather than a permanent leak.

File: src/rest.rs [hunk 9]
```diff
@@ -735,6 +764,102 @@ async fn handle_block_template_request(
...
+            let lookup = query.lookup_asset(&asset_id).await?;
+            let degraded = matches!(
+                &lookup.registry_status,
+                AssetRegistryStatus::Unavailable(_)
+            );
+            let asset_entry = lookup
+                .asset
+                .ok_or_else(|| HttpError::not_found("Asset id not found".to_string()))?;
+
+            let mut response = json_response_no_store(asset_entry, StatusCode::OK)?;
```
Comment (EddieHouston) [inline src/rest.rs]: **A registry outage returns `200` with the metadata silently missing**

When `lookup_asset` yields `AssetRegistryStatus::Unavailable` (registry 5xx, timeout, or the 16-permit semaphore rejecting the call), `IssuedAsset.meta` is `None` and this still returns `200`. Because `AssetMeta` is flattened into the response, `name`, `ticker`, `precision` and `contract` just disappear from the JSON body; the only signal is the non-standard `X-Asset-Registry-Status` header, which no existing client reads.

The practical failure is `precision`: a consumer that defaults a missing `precision` to 0 renders a 1.5-unit asset as `150000000`. With the previous on-disk registry the field was always present, so a registry blip now produces confidently wrong amounts instead of an error.

Suggested: return `502`/`503` when metadata was requested but unavailable, or at minimum keep the response shape stable with an explicit in-body status field so clients can distinguish "unregistered asset" from "registry unreachable".

File: src/elements/registry.rs [hunk 10]
```diff
@@ -1,123 +1,643 @@
...
+                    icon.href
+                )));
+            }
+            icon.href = href.to_string();
+        }
+        Ok(())
+    }
+
+    fn try_acquire_permit(&self) -> std::result::Result<OwnedSemaphorePermit, RegistryError> {
+        self.concurrency.clone().try_acquire_owned().map_err(|_| {
```
Comment (EddieHouston) [inline src/elements/registry.rs]: **16 permits with `try_acquire` means the 17th concurrent lookup is rejected outright**

`try_acquire_owned` rejects rather than queues, so with `DEFAULT_REGISTRY_MAX_CONCURRENCY = 16` the 17th in-flight distinct-asset lookup gets `RegistryError::Overloaded` immediately. That is a routine level of concurrency for this endpoint on a public instance, and with the 1s cache TTL forcing a refetch every second it is reachable at a fairly modest steady request rate.

The consequences are inconsistent: `/asset/:id` degrades to a `200` with metadata stripped (see the other comment), while `/assets/registry` and `/asset/:id/supply/decimal` return `503`. So metadata disappears non-deterministically under ordinary load.

Suggested: `acquire_owned()` under the existing request timeout so callers queue instead of failing, and/or raise the limit and make it configurable.

File: src/elements/registry.rs [hunk 11]
```diff
@@ -1,123 +1,643 @@
...
-        self.assets_cache
-            .get(asset_id)
-            .map(|(_, metadata)| metadata)
+    async fn complete_asset_fetch(&self, asset_id: AssetId, result: RegistryAssetResult) {
+        let mut cache = self.asset_cache.lock().await;
+        let waiters = match cache.remove(&asset_id) {
+            Some(AssetCacheEntry::Fetching(waiters)) => waiters,
+            _ => vec![],
+        };
+        if let Ok(asset) = &result {
```
Comment (EddieHouston) [inline src/elements/registry.rs]: **Failures are never cached, so a slow registry becomes a permanent thundering herd**

`if let Ok(asset) = &result` skips the cache insert on error, so the next `get_asset` for the same id starts a fresh fetch. Combined with `REGISTRY_ASSET_CACHE_TTL = 1s` and no negative caching, a down or slow registry means every request re-attempts upstream and pays up to the full 5s `DEFAULT_REGISTRY_REQUEST_TIMEOUT`, with 16 fetches permanently in flight.

Related: there is no stale-while-revalidate. An expired `Ready` entry is removed before the refetch starts, so concurrent callers block on the network rather than being served a value that is one second old.

Suggested: cache negative results with a short TTL, and serve the stale entry while a refresh is in flight.

File: src/elements/registry.rs [hunk 12]
```diff
@@ -1,123 +1,643 @@
...
-                continue;
+        if limit == 0 || start_index >= total_count {
+            return Ok(RegistryAssetList {
+                total_count,
+                items: vec![],
+            });
+        }
+
+        let mut items = first.items;
+        if offset.saturating_add(limit) > items.len()
```
Comment (EddieHouston) [inline src/elements/registry.rs]: **The two-page merge misaligns when the first page comes back short**

If the first page returns fewer items than `page_size`, `items.len().saturating_sub(offset)` underflows to 0 and this guard collapses to `start_index < total_count`, which is true — so a second page is fetched and appended, and the trailing `.skip(offset)` then indexes into the second page's items.

Concretely with `page_size = 25`, `start_index = 35` (page 2, offset 10): if the registry returns 5 items for page 2 (items filtered or deleted between the count and the page fetch), page 3 is appended and `.skip(10)` discards all 5 page-2 items plus the first 5 of page 3, returning a window that corresponds to no coherent `start_index`. Items are silently skipped rather than the inconsistency surfacing as an error.

Suggested: treat a short page as end-of-results (return what is available) rather than assuming `items.len() >= page_size` when computing the skip.

File: src/rest.rs [hunk 13]
```diff
@@ -1527,6 +1634,23 @@ impl HttpError {
     fn forbidden(msg: String) -> Self {
         HttpError(StatusCode::FORBIDDEN, msg)
     }
+
+    #[cfg(feature = "liquid")]
+    fn from_registry_error(error: RegistryError) -> Self {
```
Comment (EddieHouston) [inline src/rest.rs:1685]: **The raw registry error goes into the public response body, leaking the registry URL**

`RegistryError::Transport` wraps `reqwest`'s error `Display`, which includes the request URL. `from_registry_error` puts that string straight into `HttpError`, and `spawn_conn` writes it verbatim as the response body — so a caller of `/assets/registry` learns the registry's internal hostname, port and path, including any credentials embedded in `--asset-registry-url`.

Same URL is also logged unredacted via `log::debug!("configuration: {:?}", config)` in `src/config.rs`, which is the case the `SensitiveAuth` redaction added in f591f76 was meant to cover for the RPC password.

Suggested: return a fixed message ("asset registry unavailable") to the client, log the detail server-side, and give the registry URL the same `SensitiveAuth`-style redaction in the `Debug` output.

File: src/new_index/query.rs [hunk 14]
```diff
@@ -299,31 +317,116 @@ impl Query {
...
+            move || -> std::result::Result<Vec<LiquidAsset>, RegistryError> {
+                let mut results = Vec::with_capacity(items.len());
+                for registry_asset in items {
+                    let asset_id = registry_asset.asset_id;
+                    let metadata = AssetMeta::from_registry_asset(registry_asset)?;
+                    match lookup_asset(&query, &asset_id, Some(metadata))
+                        .map_err(|error| RegistryError::LocalLookup(error.to_string()))?
+                    {
+                        Some(asset) => results.push(asset),
+                        None => warn!(
```
Comment (EddieHouston) [inline src/new_index/query.rs]: **Assets missing from the local index are dropped, but `X-Total-Results` still reports the registry's count**

The previous code returned `Err("missing registered asset")`; this now logs a warning and drops the entry. The handler still sets `X-Total-Results` from the registry's `total_count`, so a page can come back with 22 items where the client expected 25, with no way to tell which are missing or that anything was dropped.

A client paginating by `start_index += limit` using `X-Total-Results` then has cursor arithmetic that silently skips real assets on every subsequent page. This is reachable normally: a freshly issued asset is in the registry before it is indexed locally, and it also happens transiently after a reorg.

Suggested: keep the count consistent with what is actually returned, or return a placeholder entry so the page length matches the requested window.

File: src/config.rs [hunk 15]
```diff
@@ -356,9 +358,9 @@ impl Config {
                     .takes_value(true),
             )
             .arg(
-                Arg::with_name("asset_db_path")
-                    .long("asset-db-path")
-                    .help("Directory for liquid/elements asset db")
+                Arg::with_name("asset_registry_url")
```
Comment (EddieHouston) [inline src/config.rs:390]: **`--asset-db-path` is removed with no alias or deprecation, so existing command lines hard-fail at startup**

Any deployment still passing `--asset-db-path /data/asset-registry-db` (including the docker-compose harness) now hits a clap "unknown argument" error and exits before indexing — a metadata-source change turns into a hard startup failure.

Also worth a migration note in the README: omitting `--asset-registry-url` silently yields empty `/assets/registry` results and metadata-free `/asset/:id` responses rather than an error, so an operator who simply drops the old flag gets a quietly degraded API instead of a clear signal.

Suggested: accept and ignore `--asset-db-path` for one release with a deprecation warning.

File: src/rest.rs [hunk 16]
```diff
@@ -735,6 +764,102 @@ async fn handle_block_template_request(
...
+            let asset_entry = lookup
+                .asset
+                .ok_or_else(|| HttpError::not_found("Asset id not found".to_string()))?;
+            let supply = asset_entry
+                .supply()
+                .ok_or_else(|| HttpError::from("Asset supply is blinded".to_string()))?;
+            let precision = asset_entry.precision();
+
+            if precision > 0 {
+                http_message(
```
Comment (EddieHouston) [inline src/rest.rs]: **`/asset/:id` is now `no-store` while `/asset/:id/supply/decimal` still caches for 10s**

`/asset/:id` went from `public, max-age=10` to `no-store`, so one of the hottest Liquid endpoints loses all CDN/reverse-proxy caching: every hit reaches electrs, and with the 1s registry cache TTL each distinct asset drives roughly one upstream registry request per second.

Meanwhile this endpoint depends on the same registry-supplied `precision` and still returns `max-age=10`, so a proxy can serve a decimal supply computed from a precision the uncacheable `/asset/:id` response no longer agrees with — and if the registry was unavailable when it was computed, the cached body is the raw integer.

Suggested: pick one policy for both. A short `max-age` on both (with the unavailable case not cached) keeps them consistent and preserves the CDN offload.

File: src/config.rs [hunk 17]
```diff
@@ -397,7 +399,14 @@ impl Config {
             });
 
         #[cfg(feature = "liquid")]
-        let asset_db_path = m.value_of("asset_db_path").map(PathBuf::from);
+        let asset_registry_url = m.value_of("asset_registry_url").map(|value| {
+            let url = Url::parse(value).expect("invalid asset registry URL");
```
Comment (EddieHouston) [inline src/config.rs]: **An invalid `--asset-registry-url` panics instead of producing a config error**

`--asset-registry-url registry.example.com/api` (no scheme) makes `Url::parse` return `RelativeUrlWithoutBase`, and this `.expect` panics with a Rust backtrace rather than a usage message; `ftp://…` trips the `assert!` just below with a second panic.

The same http/https scheme check also exists in `RegistryClient::with_options` (`src/elements/registry.rs`), so one of the two is dead validation. Suggested: keep the check in one place and return a configuration error rather than unwinding.

File: tests/rest.rs [hunk 18]
```diff
@@ -1478,6 +1633,175 @@ fn test_rest_liquid_unblinded_issuance() -> Result<()> {
     Ok(())
 }
 
+#[cfg(feature = "liquid")]
+#[test]
+fn test_rest_liquid_v2_asset_registry() -> Result<()> {
```
Comment (EddieHouston) [inline tests/rest.rs:1752]: **This test is timing-dependent and can hang the test binary**

Two issues:

1. It assumes the 1s registry cache stays warm across five REST round-trips, and hard-codes the mock to accept exactly 4 connections. On a loaded CI box, if more than 1s elapses before the later `GET /asset/:id/supply/decimal`, the entry expires and electrs opens a 5th connection — but the mock's accept loop has already ended, so the request is refused, `precision` falls back to 0, and the assertion sees a `503` instead of `"1.5"`.

2. If an assertion fails early and fewer than 4 requests reach the mock, `registry_thread.join().unwrap()` blocks on `listener.accept()` forever with no timeout, hanging the whole binary rather than reporting the failure.

Suggested: let the mock serve connections until it is signalled to stop (and set a read/accept timeout) instead of counting them, so the test is insensitive to cache timing and always terminates.

**Review by EddieHouston — COMMENTED** [review]
Automated GPT 5.6-sol re-review of this PR (medium depth). 15 findings, inline below. 
Summary of blocking ones to address before merge:

**Blocking**

1. `registry.rs` — a `Fetching` cache entry whose spawned task dies is never evicted and `receiver.await` has no timeout, so that asset id is poisoned for the process lifetime and every later request for it hangs a connection.
2. `rest.rs` — a registry outage now returns `200` with `precision`/`name`/`ticker` silently absent. A consumer defaulting a missing `precision` to 0 shows 1.5 units as `150000000`; the old on-disk path could not produce that.
3. `registry.rs` — 16 permits with `try_acquire` (reject, no queue) plus a 1s TTL and no negative caching: metadata drops out non-deterministically under ordinary concurrency, and a slow registry becomes a permanent 16-in-flight herd with each request paying the full 5s timeout.

**Worth fixing before merge**

4. The raw `RegistryError` string — including the registry URL and any embedded credentials — is written into the public response body.
5. `--asset-db-path` is deleted with no alias or deprecation, so existing command lines fail at startup.

The rest are a pagination edge case, the `no-store` vs `max-age=10` split between `/asset/:id` and its supply sibling, a panic on invalid config, a test that can hang the binary, and cleanup items.

File: src/elements/registry.rs [hunk 19]
```diff
@@ -1,123 +1,643 @@
...
+    pub async fn get_asset(
+        &self,
+        asset_id: &AssetId,
+    ) -> std::result::Result<Option<RegistryAsset>, RegistryError> {
+        let mut cache = self.asset_cache.lock().await;
+        let now = Instant::now();
+
+        if let Some(AssetCacheEntry::Ready { fetched_at, asset }) = cache.get(asset_id) {
+            if now.duration_since(*fetched_at) < self.asset_cache_ttl {
+                return Ok(asset.as_ref().map(|asset| asset.as_ref().clone()));
```
Comment (EddieHouston) [inline src/elements/registry.rs]: **Every cache hit deep-clones the asset, defeating the `Arc` it is stored behind**

`asset.as_ref().clone()` copies the whole `RegistryAsset`, including several `serde_json` maps (contract `extra`, `mutable`, `admin`, icon `extra`, top-level `extra`). `AssetMeta::from_registry_asset` then clones `contract`/`entity`/`name`/`ticker` again and re-serializes the contract to a `JsonValue` — several allocation-heavy copies per request for data that never changes. Returning `Arc<RegistryAsset>` (or caching the derived `AssetMeta`) avoids all of it.

Separately, `prune_asset_cache` runs an O(n) `retain` over up to 1024 entries plus a `min_by_key` scan per eviction, all while holding the single mutex that serialises every asset lookup in the process.

File: src/rest.rs [hunk 20]
```diff
@@ -678,14 +690,27 @@ impl Handle {
...
+/// Whether `uri` addresses the block template endpoint, one of the routes handled on the async
 /// runtime rather than on the blocking pool (see `handle_request`). Matched exactly the way
 /// the router below matches it, so the two cannot drift apart.
 fn is_block_template_request(method: &Method, uri: &hyper::Uri) -> bool {
     let mut path = uri.path().split('/').skip(1);
     *method == Method::GET && path.next() == Some("block-template") && path.next().is_none()
 }
 
+#[cfg(feature = "liquid")]
+fn is_asset_registry_request(method: &Method, uri: &hyper::Uri) -> bool {
```
Comment (EddieHouston) [inline src/rest.rs]: **The classifier is now the only matcher for these three routes**

The doc comment on `is_block_template_request` says the route is "matched exactly the way the router below matches it, so the two cannot drift apart" — which held because the arm still existed in `handle_blocking_request`. Here the `/asset/:id`, `/asset/:id/supply/decimal` and `/assets/registry` arms were deleted from the blocking router and now exist only inside `handle_asset_registry_request`, so the invariant no longer holds: if someone adds a route variant to the handler's `match path.as_slice()` but not to this classifier, the request silently falls through to the blocking router and 404s, with no compile error.

Worth generalising the mechanism (one router that reports whether a match is async) rather than growing a hand-maintained list. Minor, related: `handle_asset_registry_request` parses `uri.path()` a second time and builds both a `Vec<(String, String)>` and a full clone of it as a `HashMap` for every request.

File: src/elements/registry.rs [hunk 21]
```diff
@@ -128,185 +648,838 @@ pub struct AssetMeta {
...
-        let sort_fn: Box<dyn Fn(&AssetEntry, &AssetEntry) -> cmp::Ordering> = match self.0 {
-            AssetSortField::Name => {
-                // Order by name first, use asset id as a tie breaker. the other sorting fields
-                // don't require this because they're guaranteed to be unique.
-                Box::new(|a, b| lc_cmp(&a.1.name, &b.1.name).then_with(|| a.0.cmp(b.0)))
-            }
-            AssetSortField::Domain => Box::new(|a, b| a.1.domain().cmp(&b.1.domain())),
-            AssetSortField::Ticker => Box::new(|a, b| lc_cmp_opt(&a.1.ticker, &b.1.ticker)),
+impl AssetSearchFilters {
+    pub fn new(
```
Comment (EddieHouston) [inline src/elements/registry.rs]: **Dead code: `AssetSearchFilters::new` is never called**

`git grep AssetSearchFilters::new` finds only this definition — the REST handler uses `from_query_pairs` exclusively. Because the type is `pub` and re-exported from `src/elements/mod.rs`, the compiler emits no `dead_code` warning, so an unused two-argument constructor quietly becomes part of the crate's public surface.

Same category: `AssetMeta` still derives `Deserialize` although nothing deserializes it now that the on-disk registry is gone.

File: src/elements/registry.rs [hunk 22]
```diff
@@ -1,123 +1,643 @@
...
+            let permit = self.try_acquire_permit()?;
+            let (sender, receiver) = oneshot::channel();
+            cache.insert(*asset_id, AssetCacheEntry::Fetching(vec![sender]));
+
+            let client = self.clone();
+            let asset_id = *asset_id;
+            tokio::spawn(async move {
+                let result = client.fetch_asset(&asset_id).await;
+                if let Err(error) = &result {
+                    warn!("asset registry lookup failed for {}: {}", asset_id, error);
```
Comment (EddieHouston) [inline src/elements/registry.rs]: Nit: `asset_id` and `error` are interpolated bare here. The repo elsewhere uses the greppable `var='{}'` form (e.g. `rest.rs:718` `path='{}'`, `mempool.rs:338` `txid='{}'`), which also makes an empty value visible in logs. Same for the new `warn!` in `src/new_index/query.rs`.

File: src/elements/registry.rs (hunk 8 above)
Comment (Randy808) [reply src/elements/registry.rs]: Fixed. `AssetCacheEntry::Fetching` now carries `started_at`; both `get_asset` (before joining) and `prune_asset_cache` treat entries older than `REGISTRY_FETCH_POISON_TIMEOUT` (30s) as dead and refuse to join / evict them. `receiver.await` is also wrapped in `tokio::time::timeout(fetch_wait_timeout, …)` so a lost notification degrades to one `RegistryError::Timeout` rather than a permanent hang.

File: src/rest.rs (hunk 9 above)
Comment (Randy808) [reply src/rest.rs]: Kept. The same shape occurs under the old on-disk registry whenever the file was absent or a sync failed before ever populating an asset — the response returns 200 with a `None` meta and the metadata fields drop out. What changes here is frequency, not the failure mode, and clients can detect it via `X-Asset-Registry-Status: unavailable`. The transient-frequency problem is what the negative-cache + queue-with-timeout changes below are targeting; if we later want a stricter contract we can revisit as a follow-up.

File: src/elements/registry.rs (hunk 10 above)
Comment (Randy808) [reply src/elements/registry.rs]: Fixed. `try_acquire_permit` was replaced with `async fn acquire_permit`, which uses `tokio::time::timeout(permit_acquire_timeout, semaphore.acquire_owned())`. The timeout defaults to `request_timeout` (so the wall-clock budget matches what the caller was going to spend on the request itself) and is now a `with_options` argument, which the tests use to configure a shorter value.

File: src/elements/registry.rs (hunk 11 above)
Comment (Randy808) [reply src/elements/registry.rs]: Fixed for negative caching: `AssetCacheEntry::Ready` now holds `RegistryAssetResult`, and `complete_asset_fetch` inserts both successes and errors under the same TTL. `Fetching` de-dupes concurrent misses; a Ready-error entry short-circuits subsequent callers for one TTL. I did not implement stale-while-revalidate — that's a larger restructure and worth its own PR.

File: src/elements/registry.rs (hunk 12 above)
Comment (Randy808) [reply src/elements/registry.rs]: Fixed. Added an `items.len() == page_size` guard: a short first page is treated as end-of-results, so `.skip(offset)` no longer indexes into unrelated rows.

File: src/rest.rs (hunk 13 above)
Comment (Randy808) [reply src/rest.rs:1685]: Fixed. `HttpError::from_registry_error` returns the fixed body `"asset registry unavailable"` for anything wrapping a reqwest error (`Transport | InvalidBaseUrl | InvalidResponse | Timeout`); the raw error is logged via `warn!`. Added `config::SensitiveUrl` with a `Debug` that strips userinfo, and switched `Config.asset_registry_url` to it so the `configuration: {:?}` line no longer prints credentials.

File: src/new_index/query.rs (hunk 14 above)
Comment (Randy808) [reply src/new_index/query.rs]: Partially fixed. `Query::list_registry_assets` now returns `total_count.saturating_sub(dropped_this_page)` so `X-Total-Results` reflects what the response actually contains. Full "keep the count consistent globally" would need cross-page bookkeeping, which we can revisit if this partial fix isn't enough.

File: src/config.rs (hunk 15 above)
Comment (Randy808) [reply src/config.rs:390]: Fixed. `--asset-db-path` is accepted as a hidden clap arg; when present it prints a deprecation warning to stderr (before logging is initialised) and is otherwise ignored, so existing command lines no longer break at startup.

File: src/rest.rs (hunk 16 above)
Comment (Randy808) [reply src/rest.rs]: Fixed. `/asset/:id` now uses `Cache-Control: public, max-age=TTL_SHORT` when the registry response was healthy and `no-store` when it was degraded (matching the `X-Asset-Registry-Status: unavailable` header). `/supply/decimal` was already `TTL_SHORT` and stays that way. Degraded responses are never cached by proxies.

File: src/config.rs (hunk 17 above)
Comment (Randy808) [reply src/config.rs]: Fixed. `Config::from_args` now runs `parse_asset_registry_url`, which returns a `String` error; on error we call `clap::Error::with_description(...).exit()`, so `ftp://…` and unscheme'd values produce a clean CLI error instead of a Rust panic + backtrace. The duplicate scheme check remains only in `RegistryClient::with_options`; the CLI-side check gives the friendlier message.

File: tests/rest.rs (hunk 18 above)
Comment (Randy808) [reply tests/rest.rs:1752]: Fixed. `start_asset_registry_mock` no longer takes an expected-request count. It returns an `AssetRegistryMock` struct with a `stop()` method; the mock listens on a non-blocking socket, has 2s read/write timeouts, and serves until `stop` is signalled (also stopped on `Drop`). `test_rest_liquid_v2_asset_registry` now calls `mock.stop()` at the end and asserts `request_count >= 4` instead of exactly 4.

File: src/elements/registry.rs (hunk 19 above)
Comment (Randy808) [reply src/elements/registry.rs]: Fixed. `get_asset` returns `Option<Arc<RegistryAsset>>` and `AssetMeta.registry` is now `Arc<RegistryAsset>` (with a small `serialize_with` helper so the JSON output is unchanged). Cache hits are now an `Arc::clone` plus a small serde-json serialize of the contract; no more full deep-copy of the JSON maps. I did not change `prune_asset_cache` — the O(n) retain-under-mutex is a real concern but at 1024-entry ceiling it's tolerable; happy to revisit if it shows up on a flamegraph.

File: src/rest.rs (hunk 20 above)
Comment (Randy808) [reply src/rest.rs]: Fixed by unifying. Removed `is_asset_registry_request`; added `classify_asset_registry_request` which returns `Option<AssetRegistryRoute>` (an enum). `handle_asset_registry_request` takes the enum and matches on `(route, path)`; adding a new route now forces edits in one place, and the tuple-match keeps the enum-variant and path-shape in sync (the fallthrough is `unreachable!` because `classify_...` is the single source of truth). The duplicate parse of `uri.path()` is still there — I left it because the caller has already turned it into a `Vec<&str>` and re-parsing is cheap.

File: src/elements/registry.rs (hunk 21 above)
Comment (Randy808) [reply src/elements/registry.rs]: Fixed. Removed `AssetSearchFilters::new` and dropped the `Deserialize` derive on `AssetMeta`.

File: src/elements/registry.rs (hunk 22 above)
Comment (Randy808) [reply src/elements/registry.rs]: Fixed. `registry.rs` and `query.rs` now use the `var='{}'` convention (`asset_id='…' error='…'`).
